### PR TITLE
Add conditional spy type by the return value [WIP]

### PR DESCRIPTION
**Pull request by Bnaya**
Use ts 2.8 conditional types to expose propper types for the spys

### CODE REVIEW

File: src/spy-types.ts [hunk 1]
```diff
@@ -1,25 +1,61 @@
...
+  nextWith(value: T): void;
   nextWithError(value: any): void;
-  resolveWith(value: any): void;
-  rejectWith(value: any): void;
-}
\ No newline at end of file
+}
+
+export type AddSpyOnFunction<T extends (...args: any[]) => R, R> = T & {
+  and: jasmine.Spy;
```
Comment (Bnaya) [inline src/spy-types.ts]: The jasmine types actually a bit poor.  they don't use type information that they can get from the function (args/return value.)
We can add our own types for function spy and make the function return type more safe

File: src/playwiththenewspy.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import { Observable } from "rxjs";
```
Comment (Bnaya) [inline src/playwiththenewspy.ts]: @shairez here you can see the results of the proposed change

**Comment by shairez** [comment]
First of all, Thanks @Bnaya !

I've tested it and it looks great for promises and Observables, but we lost the regular spies now.

(If you run `npm test` you can see the error in line 30 on `create-spy-from-class.spec.ts`

Can you do more magic and make the tests pass please? 

Thanks!

**Comment by Bnaya** [comment]
I've updated the code, now the tests passing.
Note that until angular toolchain won't support 2.8 this PR is not very usefull

**Comment by shairez** [comment]
Thanks @Bnaya for the update !
Question - will it work with older versions of TypeScript? or will throw an error?

**Comment by Bnaya** [comment]
I didn't check it myself, i guess it will break older TS versions, as they don't know the new syntax.
TS 2.8 support is scheduled for angular 6.1
https://github.com/angular/angular/issues/23092

**Comment by shairez** [comment]
Thanks again @Bnaya !
I merged it, added a few changes and bumped the version number

**Pull request merged**